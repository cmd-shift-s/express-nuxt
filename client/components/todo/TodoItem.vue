<template lang="html">
  <li :class="cx">
    <div class="view">
      <input type="checkbox" class="toggle" v-model="todo.done" @change="save()">
      <label @dblclick="toggle(true)">{{item.text}}</label>
      <button class="destroy" @click="$emit('remove', item.id)"></button>
    </div>
    <input ref="edit" type="text" class="edit" v-model="todo.text" @blur="toggle(false)" @keyup.esc="toggle(false)" @keyup.enter="save()">
  </li>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'nuxt-property-decorator'
import { Todo } from '~/models'

@Component
export default class TodoItem extends Vue {
  @Prop({required: true}) item: Todo

  todo: Todo = {
    done: false,
    text: ''
  }

  edit = false

  beforeMount() {
    Object.assign(this.todo, this.item)
  }

  get cx() {
    return {
      completed: this.item.done,
      editing: this.edit
    }
  }

  toggle(edit = false) {
    this.edit = edit
    this.todo.text = this.item.text

    if (edit) {
      this.$nextTick(() => {
        (<HTMLElement>this.$refs.edit).focus()
      })
    }
  }

  save() {
    const todo = Object.assign({}, this.item, this.todo)
    this.$emit('save', todo)
    this.toggle()
  }
}
</script>

<style lang="css">
</style>
