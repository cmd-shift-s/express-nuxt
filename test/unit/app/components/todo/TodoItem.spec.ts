import { shallowMount } from '@vue/test-utils'
import TodoItem from '~/app/components/todo/TodoItem.vue'
import { Todo } from '~/share/models'

describe('components/todo/TodoItem.vue', () => {
  it('render todo.text', () => {
    const todo: Todo = {
      text: 'Hello',
      done: false
    }
    const wrap = shallowMount(TodoItem, {
      propsData: {
        item: todo
      }
    })

    expect(wrap.text()).toMatch(todo.text)
    expect(wrap.element).toMatchSnapshot()
  })

  it('render edit mode', () => {
    const todo: Todo = {
      text: 'Hello',
      done: false
    }
    const wrap = shallowMount(TodoItem, {
      propsData: {
        item: todo
      }
    })

    wrap.find('label').trigger('dblclick')

    expect(wrap.classes()).toContain('editing')
    expect(wrap.vm.$data.edit).toBe(true)
    expect(wrap.element).toMatchSnapshot()
  })

  it('should call save', () => {
    const todo: Todo = {
      text: 'Hello',
      done: false
    }
    const wrap = shallowMount(TodoItem, {
      propsData: {
        item: todo
      }
    })

    // enable edit mode
    wrap.find('label').trigger('dblclick')

    const input = wrap.find('input.edit')
    todo.text = 'World'

    input.setValue('World')
    expect(wrap.vm.$data.todo.text).toBe(todo.text)

    input.trigger('keyup.enter')

    const emitted = wrap.emitted('save')
    expect(emitted).toHaveLength(1)
    expect(emitted[0][0]).toEqual(todo)

    // disabled edit mode
    expect(wrap.classes()).not.toContain('editing')
    expect(wrap.vm.$data.edit).toBe(false)
    expect(wrap.element).toMatchSnapshot()
  })

  it('should contain "completed" class', () => {
    const todo: Todo = {
      text: 'Hello',
      done: true
    }
    const wrap = shallowMount(TodoItem, {
      propsData: {
        item: todo
      }
    })

    expect(wrap.classes()).toContain('completed')
    expect(wrap.element).toMatchSnapshot()
  })
})
