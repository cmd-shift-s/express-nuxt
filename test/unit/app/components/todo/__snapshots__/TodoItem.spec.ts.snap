// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/todo/TodoItem.vue render edit mode 1`] = `
<li
  class="editing"
>
  <div
    class="view"
  >
    <input
      class="toggle"
      type="checkbox"
    />
     
    <label>
      Hello
    </label>
     
    <button
      class="destroy"
    />
  </div>
   
  <input
    class="edit"
    type="text"
  />
</li>
`;

exports[`components/todo/TodoItem.vue render todo.text 1`] = `
<li
  class=""
>
  <div
    class="view"
  >
    <input
      class="toggle"
      type="checkbox"
    />
     
    <label>
      Hello
    </label>
     
    <button
      class="destroy"
    />
  </div>
   
  <input
    class="edit"
    type="text"
  />
</li>
`;

exports[`components/todo/TodoItem.vue should call save 1`] = `
<li
  class=""
>
  <div
    class="view"
  >
    <input
      class="toggle"
      type="checkbox"
    />
     
    <label>
      World
    </label>
     
    <button
      class="destroy"
    />
  </div>
   
  <input
    class="edit"
    type="text"
  />
</li>
`;

exports[`components/todo/TodoItem.vue should contain "completed" class 1`] = `
<li
  class="completed"
>
  <div
    class="view"
  >
    <input
      class="toggle"
      type="checkbox"
    />
     
    <label>
      Hello
    </label>
     
    <button
      class="destroy"
    />
  </div>
   
  <input
    class="edit"
    type="text"
  />
</li>
`;
